asio/include/asio/detail/is_buffer_sequence.hpp

```diff
@@ -252,49 +92,76 @@ struct is_buffer_sequence<const_buffer, mutable_buffer>
 {
 };
 
-template <typename T>
-struct is_dynamic_buffer_class_v1
-  : integral_constant<bool,
-      sizeof(size_memfn_helper<T>(0)) != 1 &&
-      sizeof(max_size_memfn_helper<T>(0)) != 1 &&
-      sizeof(capacity_memfn_helper<T>(0)) != 1 &&
-      sizeof(data_memfn_helper<T>(0)) != 1 &&
-      sizeof(consume_memfn_helper<T>(0)) != 1 &&
-      sizeof(prepare_memfn_helper<T>(0)) != 1 &&
-      sizeof(commit_memfn_helper<T>(0)) != 1 &&
-      sizeof(const_buffers_type_typedef_helper<T>(0)) == 1 &&
-      sizeof(mutable_buffers_type_typedef_helper<T>(0)) == 1>
+struct is_dynamic_buffer_class_v1 : sfinae_check_base
 {
+  static const std::size_t arg = 0;
+
+  template <typename T>
+  static result<sizeof(
+
+    is_same_as<std::size_t>(declval<T>().size()),
```

Comment:
Perhaps we should make this check (and other similar checks) more strict:
```suggestion
    is_same_as<std::size_t>(declval<const T>().size()),
```
And also to relax the return type requirement:
```suggestion
    is_convertible_to<std::size_t>(declval<T>().size()),
```